File: src/leetCode/c++/Solution238.cpp
//O(n) O(1)
class Solution238 {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size());
        int temp = 1;
        for(int i=0; i<nums.size(); ++i){
            res[i] = temp;
            temp *= nums[i];
        }

        temp = 1;
        for(int i=nums.size()-1; i>=0; --i){
            res[i] *= temp;
            temp *= nums[i];
        }
        return res;
    }
};
/*
1  2  3  4
1  1  2  6
24 12 4  1

24 12 8  6
*/

File: java/Solution238.java
package leetCode.java;

//O(n) O(1)
class Solution238 {
    public int[] productExceptSelf(int[] nums) {
        // must run in O(n), 典型的 prefix product + suffix product (前綴積 + 後綴積)題型
        int[] res = new int[nums.length];

        int prefix = 1;
        for (int i = 0; i < nums.length; i++) {
            res[i] = prefix;
            prefix *= nums[i]; // 過了之後才乘
        }

        int suffix = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            res[i] *= suffix;
            suffix *= nums[i]; // 過了之後才乘
        }

        return res;
    }
}